Non-unreal server worker types [stage 2] (#879)
spatialos/UnrealGDK: SpatialGDK/Documentation/content/non-unreal-server-worker-types.md

Default single Unreal server-worker type development doesn’t accommodate schema files that haven’t been generated by the GDK, so you need to set your game up to handle this.

> >**TIP:** If you’re using schema from outside the GDK, you can customise [snapshot generation]({{urlRoot}}/content/spatialos-concepts/generating-a-snapshot) from the GDK toolbar’s  **Snapshot** button. This is to serialize additional entities with these external components which the default Unreal snapshot generation cannot currently do. See [Add to the snapshot](#add-to-the-snapshot) below for how to do this.

Reviewer: @ElleEss and/or @NotRobSheridan could you take a look over these documentation changes.

Reviewer 2: Looking!

* For each component, the `ExternalSchemaInterface` will contain methods for `SendComponentUpdate`, `OnAddComponent`, `OnComponentUpdate`, `OnAuthorityChange`, and `OnRemoveComponent`.
* For each command, the `ExternalSchemaInterface` will contain methods for `SendCommandRequest`, `SendCommandResponse`, `OnCommandRequest` and `OnCommandResponse`.

##### Setting up

Below is a summary of the steps required to use the external schema code generator. There is a more complete example in the [examples section below](#using-the-code-generator-example).

* Run the `ExternalSchemaCodegen.bat` file as described in the [helper scripts documentation]({{urlRoot}}/content/helper-scripts) to generate code from a directory containing external schema files.

> * Regenerate the Visual Studio solution.

Reviewer 3: ```suggestion
* [Regenerate](https://wiki.unrealengine.com/Generate_Visual_Studio_Project) the Visual Studio project files.
```

Default single Unreal server-worker type development doesn’t accommodate schema files that haven’t been generated by the GDK, so you need to set your game up to handle this.

> >**TIP:** If you’re using schema from outside the GDK, you can customise [snapshot generation](https://docs.improbable.io/unreal/alpha/content/generating-a-snapshot#generating-a-snapshot) from the GDK toolbar’s  **Snapshot** button. This is to serialize additional entities with these external components which the default Unreal snapshot generation cannot currently do. See [Add to the snapshot](#add-to-the-snapshot) below for how to do this.

Reviewer 4: Could you use `{{urlRoot}}` for the snapshot generation link to prevent it from being broken in the future?

>**Note:** Your external SpatialOS components must have an ID between 1000 and 2000 to be registered by the pipeline.

#### Using the code generation

The external schema code generation tool generates C++ code in your game module from a specific directory containing schema files. The generated code includes classes representing each external schema enumeration and type, and an `ExternalSchemaInterface` class for sending updates and registering callbacks for receiving network operations. The `ExternalSchemaInterface` generates methods for each component and command within a component in your external schema.

* For each component, the `ExternalSchemaInterface` will contain methods for `SendComponentUpdate`, `OnAddComponent`, `OnComponentUpdate`, `OnAuthorityChange`, and `OnRemoveComponent`.
* For each command, the `ExternalSchemaInterface` will contain methods for `SendCommandRequest`, `SendCommandResponse`, `OnCommandRequest` and `OnCommandResponse`.

##### Setting up

> To interact with the external schema types by using the code generation tool, you'll need to do the following:

Reviewer 4: I saw there's a detailed explanation of these steps with examples below. Do you think it's worth mentioning this here or linking for each bullet point `(see example)`?

* To send SpatialOS component updates and commands, call the overloaded `SendComponentUpdate`, `SendCommandRequest`, or `SendCommandResponse` method for the relevant schema type.
* To register callbacks for receiving SpatialOS component network operations, call the overloaded `OnAddComponent`, `OnComponentUpdate`, `OnAuthorityChange`, `OnRemoveComponent`, `OnCommandRequest` or `OnCommandResponse` method for the relevant schema type.

##### Generated Code Format

// describe type implementation here

#### Manual (de)serialization

You can also manually serialize and deserialize C API network operations. This is what generated code alternative described above does under the hood. To do this, you'll need to do the following:

To send SpatialOS component updates and commands, use methods defined in the `SpatialWorkerConnection.h` file:

* `void SendComponentUpdate(Worker_EntityId EntityId, const Worker_ComponentUpdate* ComponentUpdate);`
* `Worker_RequestId SendCommandRequest(Worker_EntityId EntityId, const Worker_CommandRequest* Request, uint32_t CommandId);`
* `void SendCommandResponse(Worker_RequestId RequestId, const Worker_CommandResponse* Response);`

> which can access via a reference to the net driver:

Reviewer 4: "can be accessed" or the previous version which uses active voice and is preferred by tech writers afaik

`SpatialWorkerConnection connection = Cast<USpatialNetDriver>(World->GetNetDriver())->Connection;`

> To receive [network operations](https://docs.improbable.io/reference/latest/shared/design/operations) for external SpatialOS components, you must provide custom callbacks for specific component IDs and operation types via the `USpatialDispatcher::AddOpCallback` functions. These overloaded functions are parameterized with a `Worker_ComponentId` and a callback function const reference that takes one of the following network operation types as an argument:

Reviewer 4: I think `AddOpCallback` is out of date - https://github.com/spatialos/UnrealGDK/blob/master/SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialDispatcher.h

Author: yep nice, this should've been done in previous PR, make separate PR now

    string some_response_string = 1;
}

component TestComponent {
    id = 1337;
    uint32 counter = 1;
    command UnrealResponse test_command(UnrealRequest);
}

component OtherTestComponent {
    id = 1338;
    uint32 other_counter = 1;
}
```

#### Using the code generator example

Using the [Third Person Shooter project](https://github.com/spatialos/UnrealGDKThirdPersonShooter) as an example, and the above schema file placed in the directory `./spatial/schema/improbable/`, the command below will output code to a new `./Game/Source/ThirdPersonShooter/ExternalSchemaCodegen` folder:

> `./Game/Plugins/UnrealGDK/SpatialGDK/Build/Scripts/ExternalSchemaCodegen.bat ./spatial/schema/my_external_schema ./Game/Source/ThirdPersonShooter/ExternalSchemaCodegen`

Reviewer 5: Change to "MyProject"

> `./Game/Plugins/UnrealGDK/SpatialGDK/Build/Scripts/ExternalSchemaCodegen.bat ./spatial/schema/my_external_schema ./Game/Source/ThirdPersonShooter/ExternalSchemaCodegen`

Reviewer 5: Change to "MyProject"

Reviewer 5: We should be using cmd style here such as `Game\Plugins\UnrealGDK\SpatialGDK\Build\Scripts\ExternalSchemaCodegen.bat spatial\schema\session Game\Source\GDKShooter\ExternalSchemaCodegen`

> This will generate code in the `Game\Source\<your_project_name>\ExternalSchemaCode` folder. To see these changes, the Visual Studio solution needs to be regenerated.

Reviewer 3: ```suggestion
This will generate code in the `Game\Source\<your_project_name>\ExternalSchemaCode` folder. To see these changes in Visual Studio, the Visual Studio project files need to be regenerated.
```

To send or receive network operations, you need to instantiate the `ExternalSchemaInterface` class inside your game module.

To send component updates or command requests / responses, you can call any of the overrides for `SendComponentUpdate`, `SendCommandRequest(...)` or `SendCommandResponse(...)`, for example:

```
auto Update = new ::improbable::testing::TestComponent::Update(20 /* counter field */);
ExternalSchema->SendComponentUpdate(30 /* entity_id */, Update);
```

> 

Reviewer 3: In addition, there seems to be generated functions for each field (SendCounterUpdate()) etc. I think we should mention at least the existence of these helper functions.

Author: The generated fields for each field are constructor pattern style methods, added a comment between these two lines explaining this:

`// further edit Update using constructor pattern methods, e.g. (void)Update->SetCounter(30);`

<br/>

<br/>------------<br/>

> 2019-06-28 Page updated with draft content: code generator, setting up, manual (de)serialzation added

Reviewer 2: ```suggestion
_2019-06-28 Page updated with draft content: code generator, setting up, manual (de)serialzation added_
```

> 2019-06-28 Page updated with draft content: code generator, setting up, manual (de)serialzation added

Reviewer 2: ```suggestion
_2019-06-28 Page updated with draft content: code generator, setting up, manual (de)serialzation added_
```

> [//]: # (https://improbableio.atlassian.net/browse/DOC-1189)

Reviewer 2: ```suggestion
[//]: # (TODO: https://improbableio.atlassian.net/browse/DOC-1189)
```
